### PR TITLE
Lib/javascript/v8: resolve race conditions in context-aware init.

### PR DESCRIPTION
**Pull request by dot-asm**
Rare crashes in initialization procedure were observed with worker_threads. Solution is to serialize it. This is [the] one prior to the last commit.

Then there is a trivial test for worker_threads in the last commit. I'm not sure if ~with~this would be an acceptable way to solve it. If not, elaborate on expectations.

Two remaining commits are unrelated to the main problem, the race condition, but I figured I'll throw them in at once...

### CODE REVIEW

**Comment by dot-asm** [comment]
Additional commit is not directly related to the main problem, race condition in initialization. I've been browsing in v8 repository and couldn't justify some of the pre-processor conditions, hence the cleanup...

File: Examples/test-suite/javascript/using2_workers_runme.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+try {
+    const { Worker } = require("worker_threads");
+
+    const worker1 = new Worker("./using2_runme.js");
+    const worker2 = new Worker("./using2_runme.js");
+
+    worker1.on('exit', (code) => { });
+    worker2.on('exit', (code) => { });
+} catch(err) {
```
Comment (wsfulton) [inline Examples/test-suite/javascript/using2_workers_runme.js:9]: This swallows all exceptions, so any problems in the test get masked for all versions of Javascript. Seems hardly worth having the runtime test! Can you not detect the supported version at runtime and then only test.

File: Examples/test-suite/javascript/Makefile.in [hunk 2]
```diff
@@ -84,7 +84,11 @@ ifeq (node,$(JSENGINE))
 
 	run_testcase = \
 		if [ -f $(srcdir)/$*$(SCRIPTSUFFIX) ]; then \
-			env NODE_PATH=$$PWD:$(srcdir) $(RUNTOOL) $(NODEJS) $(srcdir)/$*$(SCRIPTSUFFIX); \
+			export NODE_PATH=$$PWD:.; cd $(srcdir); \
+			$(RUNTOOL) $(NODEJS) $*$(SCRIPTSUFFIX); \
+			if [ -f $*_workers$(SCRIPTSUFFIX) ]; then \
+				$(RUNTOOL) $(NODEJS) $*_workers$(SCRIPTSUFFIX); \
+			fi \
```
Comment (wsfulton) [inline Examples/test-suite/javascript/Makefile.in:91]: I really don't want any complications in the makefile in the generic case and I'd rather keep the environment variable setting pattern we have using env instead of changing to export. I'd prefer any complications went into an individual runme.js file. I think I've missed something completely here, but can you not put Worker testing into another testcase that does not already have a runme.js file?

File: Examples/test-suite/javascript/using2_workers_runme.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+try {
```
Comment (wsfulton) [inline Examples/test-suite/javascript/using2_workers_runme.js:1]: Could add in a description of what this test is testing please. I'm afraid it isn't clear to me.

File: Lib/javascript/v8/javascriptcode.swg [hunk 4]
```diff
@@ -377,8 +377,10 @@ fail:
 %{
   /* Name: $jsmangledname, Type: $jsmangledtype, Dtor: $jsdtor */
   SWIGV8_FUNCTION_TEMPLATE $jsmangledname_class = SWIGV8_CreateClassTemplate("$jsmangledname");
-  SWIGV8_SET_CLASS_TEMPL($jsmangledname_clientData.class_templ, $jsmangledname_class);
-  $jsmangledname_clientData.dtor = $jsdtor;
+  if ($jsmangledname_clientData.class_templ.IsEmpty()) {
+    SWIGV8_SET_CLASS_TEMPL($jsmangledname_clientData.class_templ, $jsmangledname_class);
+    $jsmangledname_clientData.dtor = $jsdtor;
+  }
```
Comment (wsfulton) [inline Lib/javascript/v8/javascriptcode.swg:383]: This code goes into the SWIGV8_INIT init function. I think you've put this in because it is possible for the init  function to be called more than one?? If so, it feels like it needs to be thread-safe and this change would not solve a race condition. There would also need to be a ton more work done in the init function as a lot of it operates on global variables.

File: Examples/test-suite/javascript/Makefile.in (hunk 2 above)
Comment (dot-asm) [reply Examples/test-suite/javascript/Makefile.in:91]: One can enclose it in `()` to prevent environment variable "leakage". Just in case, rationale was that normal make invokes separate shell for each line, so there won't be any actual leakage to next test. Of course one might wonder if there is some make variable that would execute everything in single shell... Either way, should I add `()`? Or should I think of something completely different?

As for putting Worker test into separate testcase. Well, I still need a swig module to `require`, and it's only natural to reuse one that is already compiled. Then one needs a script to pass to `new Worker()`-s as argument, and it's only natural to reuse one that is already available. Of course one can pass a string to `new Worker()`-s and complement it with `{eval: true}`, I've played with this option, but found it somewhat ugly. Simply grabbing an existing script felt more natural and was more elegant.

Note that we can skip this commit and do it in separate PR...

File: Lib/javascript/v8/javascriptcode.swg (hunk 4 above)
Comment (dot-asm) [reply Lib/javascript/v8/javascriptcode.swg:383]: > If so, it feels like it needs to be thread-safe

That's `isolate->Enter();` and `isolate->Exit();`'s job. I mean those are responsible for serialization of complete init function, so that we can safely check for all of IsEmpty-s.

File: Examples/test-suite/javascript/Makefile.in (hunk 2 above)
Comment (dot-asm) [reply Examples/test-suite/javascript/Makefile.in:91]: > Or should I think of something completely different?

Most notably if approach to check for presence of a "side-kick" `$*_workers$(SCRIPTSUFFIX)` is generally acceptable?

File: Examples/test-suite/javascript/using2_workers_runme.js (hunk 1 above)
Comment (dot-asm) [reply Examples/test-suite/javascript/using2_workers_runme.js:9]: > Can you not detect the supported version at runtime and then only test.

The source of my hesitation to do so is that I didn't want to make it node-specific. I mean node version detection is node-specific, and I figured that it won't be appreciated, because the scripts are exercised with other javascript engines. Or do you mean detect node version in make rule? This of course can be done... Even though workers made their appearance in 10.5, it was 12.0 that shipped them as integral part of the run-time environment.

Comment (dot-asm) [reply Examples/test-suite/javascript/using2_workers_runme.js:9]: > I didn't want to make it node-specific.

I think I've conflated unrelated thoughts here. I mean given the **presented** suggestion we already know that the script in question is executed exclusively by node. Hence it doesn't have to be engine-neutral. But I did think about it initially, and apparently failed to disregard the thought when commenting. So that yes, it's totally possible to detect if we are running on node v12+ and perform test only then.

I'll wait for response to the question if "side-kick" `$*_workers$(SCRIPTSUFFIX)` is generally acceptable before I proceed.

File: Examples/test-suite/javascript/Makefile.in (hunk 2 above)
Comment (dot-asm) [reply Examples/test-suite/javascript/Makefile.in:91]: > As alternative and more extreme approach we can consider testing **all** `$*$(SCRIPTSUFFIX)`-s with workers. By first executing `$(NODEJS) $*$(SCRIPTSUFFIX)` and then `$(NODEJS)  $(srcdir)/workers.js $*$(SCRIPTSUFFIX)`, where workers.js would instantiate whichever script passed on the command line instead of hard-coding specific one...

For reference, such test with v12 results in few SEGV and unhandled exception in template_static. And is <1% slower. As initially mentioned, the suggested context-aware init is formally incomplete and needs further work. It's just an initial step, but it's expected and was shown to work with purely computational tasks. In other words some failures can be expected in general case. **When** taken into Workers that is. Which was not even a possibility so far.

Bottom line is that "side-kick" approach can be preferred as intermediate step, till remaining general-case problems are resolved.

**Comment by dot-asm** [comment]
I've re-based to current master, reworked workers testcase, and added another unrelated commit, "use ::Cast instead of To* when possible."

**Comment by dot-asm** [comment]
> reworked workers testcase

In essence, just forget what was said so far about the testcase, and have a fresh look.

**Comment by dot-asm** [comment]
Ah! node v10- has failed. Additional fix is coming in soon...

**Comment by dot-asm** [comment]
Hmmm... It looks like I've fallen victim to kind of "propaganda." When you read about context-aware init and workers you get an impression that each worker gets own `context`, and it's effectively implied that they all share same `isolate`. But if I print `isolate` [which is a pointer] in init, they turn to be all unique... This means that the problem requires some further digging and consideration. This bring us to the following question. Since this PR has some commits that are not related to the main problem, it might be sensible to separate them and proceed with the main problem in separate PR. What's most practical for you? You can simply cherry-pick commits up to "resolve race condition", or should I remove top two commits here and effectively re-purpose this PR?

**Comment by wsfulton** [comment]
Okay, I'll commit them (up to and excluding 'resolve race condition') later today.

BTW if you really need two runme tests in the test framework, I was thinking you could instead get the Worker to call the same script with one/more commandline arguments, which you then detect and run the different code you want the worker to test. The main script takes no commandline arguments. That way the same script is self contained with dual functionality and we can keep the test-framework simple.

**Comment by dot-asm** [comment]
> BTW if you really need two runme tests in the test framework

Final iteration was one additional script that can be used with arbitrary runme.js. Final goal is to execute **all** the [runnable] tests by themselves **and** with workers.

**Comment by wsfulton** [comment]
See 3f391478c481da4d1cc840b5647936dfad9b09c2.

**Comment by dot-asm** [comment]
Thanks! The rest would have to be re-classified as "making module initialization [even] isolate-aware." I count that there is no emergency releasing 4.1, as it might take me some time. But if there is pressure, then say the word, I can re-prioritize... Thanks again:-)

**Comment by wsfulton** [comment]
There is no reason to hold up 4.1 for these changes. But then it'll probably be a couple more months until 4.1 is ready.
